### DISCUSSION TITLE
Improve the flow for adding the Product Collection block

### DISCUSSION THREAD

**Discussion by shaunandrews** [woocommerce/woocommerce#45040] [WooCommerce, Blocks, and Site Editing]
Currently, when you add a new Product Collection block to the canvas it defaults to displaying a 3 column grid, ordered alphabetically, with some random stock status filters ("In stock, Out of stock, On backorder" — why?). I'm able to use the toolbar to switch views, and adjust the paging options. I need to dig into the sidebar to find some more filters and settings.

<img width="705" alt="The default state of a newly added Product Collection block. It shows a 3-column grid of products with a block toolbar." src="https://github.com/woocommerce/woocommerce-blocks/assets/191598/8ef45e19-054a-4a5f-8678-83e9b49a07fc">

One way to improve this might be to incorporate a placeholder state, similar to WordPress Core blocks like Image and Columns. Our placeholder could offer options to choose a collection, or create a new collection.

<img width="1718" alt="CleanShot 2023-08-22 at 16 53 30@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/191598/e4fbfdcf-23f3-4c1f-bebc-1a4caec6d5da">

Selecting "Choose a collection" would trigger an overlay containing a list of patterns and presets, including: New Arrivals, Featured, On Sale, and more. These patterns and presets need to be further defined.

<img width="1751" alt="CleanShot 2023-08-22 at 16 52 34@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/191598/8d99206a-61e0-4f48-ab27-c7a014bd40d8">

https://github.com/woocommerce/woocommerce-blocks/assets/191598/8fd9b551-0dfb-4f63-aef4-a12e73bd6270

I'm not entirely sure about creating a new collection yet, but my thinking is we could allow builders to add filters and other settings to define a reusable collection of products. Here's a quick example of how that might work.

https://github.com/woocommerce/woocommerce-blocks/assets/191598/31f33f28-a71e-4201-9aa7-96becc077501

My uncertainty comes from some similarities in the way WP Core's block Variations and Synced Patterns collide.

**Comment by shaunandrews** [comment]
We could also look to offer direct access to some of these variations when searching the block library:

<img width="1512" alt="CleanShot 2023-08-23 at 11 13 44@2x" src="https://github.com/woocommerce/woocommerce-blocks/assets/191598/4b5ada00-7dc8-491e-af02-f5fdd406ce27">

**Reply by imanish003** [reply]
Although I am uncertain whether we will choose WP Core's block Variations and Synced Patterns (distinct from regular patterns), it appears that both are available when searching using /name.

**Comment by kmanijak** [comment]
I did some initial research on that topic. Here's the summary of my findings.

## Product Collection: Patterns vs Collections

First, I'd like to highlight the semantic difference between:
- **Patterns** -  represent the UI aspect of Product Collection, mainly: the layout (grid vs. list), number of columns, alignment and product template composition. https://github.com/woocommerce/woocommerce-blocks/issues/9372
- **Collections** - represent the query behind the Product Collection (filters, sorting). But they also may influence UI as patterns. https://github.com/woocommerce/woocommerce-blocks/issues/10822

Both aspects - UI and query - are represented by `attributes` so technically they are pretty similar, but the usage is different and we need to take that into account during implementation. I tried to gather the requirements/use cases in which they differ:

  | Pattern | Collection | Description or image
-- | -- | -- | --
Display in "Choose pattern" modal | ✅ | ✅ |  <img width="317" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/be0d4010-ee61-4085-9fd6-13a048f37daf">
Display in "Choose a collection " | ✅ | ✅ |  <img width="259" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/42c0b269-9197-4d6d-b704-8295fba89545">
Display in Inserter as Pattern | ✅ | ❌ |  <img width="344" alt="Screen Shot 2023-09-12 at 11 47 05 AM" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/d44d6717-2964-4087-b9e1-f74081ae68b2">
Display in Inserter as Block | ❌ | ✅ |  <img width="250" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/06c7d595-a75f-4359-86d7-01fbf7edd7b1">
Ability to Add, Edit, or Remove a custom one | ❌ | ✅ |  Collections should be manageable by merchants, but that means, we have to be able to add, edit, remove Collections programmatically.

---

## Collections: Implementation options

While "UI Patterns" from https://github.com/woocommerce/woocommerce-blocks/issues/9372 are implemented as "patterns", it's no longer obvious with Collection. Below, I'm exploring three possibilities. Collections as:
- Patterns
- Synced Patterns (former Reusable Blocks) 
- Variations

### Patterns

Patterns are a convenient way of storing block compositions. In the same way, we can also store the block's `attributes`, which allows us to store query information. 

##### 👍 Pros
1. Easy way to get Product Collection's patterns via [`getPatternsByBlockTypes`](https://developer.wordpress.org/block-editor/reference-guides/data/data-core-block-editor/#getpatternsbyblocktypes) selector. Like in https://github.com/woocommerce/woocommerce-blocks/pull/10876.

##### ❌ Cons
1. There's no clear and obvious way to register/manage patterns dynamically. There's [PHP API](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-patterns/#block-patterns) but it's not designed to be used from the Editor. And I don't think adding additional overhead (e.g. new Store API method that would expose pattern registration) is a way to go.
2. Patterns are not intended to be available in the inserter as blocks.

##### Summary
The cons are pretty strong and there are technical limitations to achieve the expected outcome without big implementation overhead (and potentially hacking). **Hence I think this option is disqualified.**

---

### Synced Patterns (former Reusable Blocks)

[Synced Patterns](https://wordpress.org/news/2023/07/synced-patterns-the-evolution-of-reusable-blocks/) is a cool feature that seems to match the Collection use case. I played with them and am not sure anymore.

<img width="1633" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/311d5501-4c4e-441a-bd03-e4dd37f08d67">

##### 👍 Pros
1. Handles user actions: adding, editing, and removing Collections.
<img width="1257" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/058e04b6-c950-4bbe-a519-3ef77bd55482">

2. Collection synchronization across the whole store.
3. Collection is available in the inserter.

##### ❌ Cons
1. We need to be able to programmatically create new Collection. The actions to create/delete Synced Patterns are still experimental so that could be an obstacle relying Collections on  Synced Patterns:
    - [`__experimentalConvertBlocksToReusable`](https://github.com/WordPress/gutenberg/blob/b05bee7e2e8eefb2be06c228002e4f9783847747/packages/reusable-blocks/src/store/actions.js#L49)
    - [`__experimentalDeleteReusableBlock`](https://github.com/WordPress/gutenberg/blob/b05bee7e2e8eefb2be06c228002e4f9783847747/packages/reusable-blocks/src/store/actions.js#L95C14-L95C47)
However, it requires confirmation, as there are no major changes for the last 2 years and Synced Patterns rely on this code.
2. Synced Pattern is in fact "a virtual" block that wraps the content. Turning Product Collection into Synced Pattern makes it a named Synced Pattern. But that means Collection is no longer represented as Product Collection instance. All Synced Patterns can be get by `select('core/edit-site').getReusableBlocks()` but there's no obvious data that would allow us to distinguish Collection from any other Synced Pattern except for parsing raw content and looking for Product Collection:
<img width="417" alt="Screen Shot 2023-09-12 at 14 53 14 PM" src="https://github.com/woocommerce/woocommerce-blocks/assets/20098064/87df74f0-6bb4-4b51-a577-1b9315c98841">

Alternatively, we could establish a naming convention, for example enforcing a prefix like "Product Collection - XYZ". But I don't consider it a good UX.

##### Summary:

Synced Patterns offer a lot in terms of what we want from Collections (flexibility, easy management) but at the same time, they are abstract entities which makes them hard to use in specific contexts.

---

### Variations

It's a well-established [API](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-variations/) that may serve the purpose of Collections. It has a clear way of registering and unregistering patterns. We can easily associate variations with the Product Collection block, and add custom names, descriptions, and even icons - all out of the box.

##### 👍 Pros
1. Mature API.
2. Huge flexibility in terms of name, icon, and description.
3. Clear association of the Collections with Product Collection block and easy to get all block variations by [`getblockvariations`](https://developer.wordpress.org/block-editor/reference-guides/data/data-core-blocks/#getblockvariations).
4. Variations can be available in the inserter out of the box.

##### ❌ Cons
1. The biggest problem I see is the persistence of the custom variations. Variations are not stored between sessions, meaning that each time the merchant starts a new session, a variation has to be registered. That's ok with predefined collections (we know the `attributes` and `innerBlocks` and they are static). With custom ones, we need to figure out the way we'd like to store our variations. Is it simply DB or there's a better way of doing that?
2. I was able to achieve a preview of Collection among Patterns using `__experimentalBlockPatternsList` and adjusting Variation structure to Pattern structure. However, that may be considered hacking, so I'm mentioning this in cons.

##### Summary: 
We know the Variation API has been overused before (https://github.com/woocommerce/woocommerce-blocks/discussions/9010), but based on my research, seems like Variations API matches our needs the best. And seems like Collections would actually be a variation of Product Collection - they have different layouts and query attributes, but underneath they're Product Collection with the same core functionality.

I made a POC (https://github.com/woocommerce/woocommerce-blocks/pull/10933) in which I registered sample variation (New Arrivals) with custom attributes and inner blocks structure. Variations are adjusted so they can be displayed with a preview among patterns. There are also two strategies:
- If you choose collection the whole block is replaced (UI and query)
- If you choose a pattern, only UI aspect is overridden.

---

## Collections: My recommendation

Based on the research, I believe we should go with Variations. Varations allow for the most of our needs and have the most mature API that we can rely on. There's still a need for custom solution to store custom collections created by the user, but I don't think that's a blocker.

---

## Next steps

Based on the POC (https://github.com/woocommerce/woocommerce-blocks/pull/10933) where you can see a short demo of:
- Predefined collection and how it behaves comparing to "blank" Product Collection.
- Collection in the Pattern/Collection inserter.

I'm going to start creating a Placeholder view with the option to choose Collection or create a new one with the Variation API as a way to go, hovewer, it's an open discussion and if you think we should go another way or there's another option not explored in this post, please share your opinion! 🙌 

cc: @imanish003, @tjcafferkey, @nerrad

**Reply by imanish003** [reply]
Hey Karol, thank you for the detailed breakdown of your research – it's evident that a lot of thought has gone into this 🚀. Here are my inputs and questions that I have:

> Display in Inserter as Pattern  

I'm not certain if hiding the "Collection" option from the inserter is necessary, unless it's technically infeasible to implement.

> There's no clear and obvious way to register/manage patterns dynamically. There's [PHP API](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-patterns/#block-patterns) but it's not designed to be used from the Editor. And I don't think adding additional overhead (e.g. new Store API method that would expose pattern registration) is a way to go.  

You mentioned that the [PHP API](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-patterns/#block-patterns) isn't conventionally used from the editor, could you elaborate more on the drawbacks and why it is deemed unfit for our requirements?

> However, it requires confirmation, as there are no major changes for the last 2 years and Synced Patterns rely on this code.  

Seeking confirmation on the usage of experimental features indeed seems like a pivotal step. Given that there have been no substantial changes in Synced Patterns over the past two years, determining its stability and projected longevity is crucial. Could we possibly reach out to someone familiar with the development trajectory of this feature to gain insights?

> Alternatively, we could establish a naming convention, for example enforcing a prefix like "Product Collection - XYZ". But I don't consider it a good UX.  

I share your concerns regarding the establishment of a naming convention. It doesn't quite align with the goal of delivering a seamless UX. Perhaps @shaunandrews could weigh in with some UX expertise on whether this route could have merits?

#### Closing Thoughts

It seems that both Block Variations and Synced Patterns are strong contenders:
- They both leverage experimental APIs:
    - For patterns: `__experimentalDeleteReusableBlock & __experimentalConvertBlocksToReusable`
    - For block variations: `__experimentalBlockPatternsList`
- The challenge will be determining which of these experimental APIs offers more stability and is future-proof.

In terms of distinguishing collections from other synced patterns, as per your findings, the options we have are either:
1. Adopting a prefix or suffix in the title, or
2. Processing the raw value.

Neither of these solutions seems ideal, echoing your sentiments.

At this stage, proceeding with Block Variations seems to be the more sensible choice given that collections essentially represent variations of the Product Collection block.

**Reply by kmanijak** [reply]
Thanks for the response @imanish003 ! 🙌 

---

> > Display in Inserter as Pattern
> I'm not certain if hiding the "Collection" option from the inserter is necessary, unless it's technically infeasible to implement.

I'd like to highlight the difference between displaying Collection in inserter _as Pattern_ (preview) vs. _as block_ (block icon). Collection should be displayed as a block, but not necessarily as a pattern. Does it make sense?

---

> You mentioned that the [PHP API](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-patterns/#block-patterns) isn't conventionally used from the editor, could you elaborate more on the drawbacks and why it is deemed unfit for our requirements?

We don't have direct access to this API on the client-side so that would require exposing it. But also, I'm not sure if the API is designed to be used dynamically. Following the notes in the [documentation](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-patterns/#block-patterns): 

> register_block_pattern() should be called from a handler attached to the init hook.

And in general, following this path seems like a hustle.

---

> Seeking confirmation on the usage of experimental features indeed seems like a pivotal step. Given that there have been no substantial changes in Synced Patterns over the past two years, determining its stability and projected longevity is crucial. Could we possibly reach out to someone familiar with the development trajectory of this feature to gain insights?

We could, but the question is if it's worth it to seek the answer knowing that the rest of cons won't allow us to use this solution (I mean lack of clear association between Synced Pattern and particular block). If you think that positive answer about the stability of this API may influence the final decision, let me know and I'll reach out for answers. Otherwise, I don't think it's worth it.

---

> At this stage, proceeding with Block Variations seems to be the more sensible choice given that collections essentially represent variations of the Product Collection block.

💪 Happy we have the same vision, thanks @imanish003!

**Reply by imanish003** [reply]
> Collection should be displayed as a block, but not necessarily as a pattern. Does it make sense?  

It certainly makes sense if we want to keep Collection separate from Pattern, and it seems like this is our direction, as discussed in Slack( p1694700105645609-slack-C02FL3X7KR6 ).

> We don't have direct access to this API on the client-side so that would require exposing it. But also, I'm not sure if the API is designed to be used dynamically. Following the notes in the [documentation](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-patterns/#block-patterns)

Got it. Thanks 🤝

> We could, but the question is if it's worth it to seek the answer knowing that the rest of cons won't allow us to use this solution (I mean lack of clear association between Synced Pattern and particular block). If you think that positive answer about the stability of this API may influence the final decision, let me know and I'll reach out for answers. Otherwise, I don't think it's worth it.  

Agreed! Given that the block variations appear to fulfill our current requirements adeptly, it doesn't seem necessary to delve deeper into this. However, it could be worth revisiting if others perceive a potential value in exploring this further. The important thing is that we choose the route that ensures stability and meets our needs efficiently 💪🏻

Thank you once again for the thorough research 🚀
